### ISSUE TITLE
Kotlin ad-hoc tasks should be up-to-date when script changes but inputs and outputs do not

### ISSUE THREAD

**Issue by mkobit** [gradle/kotlin-dsl#1271]
Editing a build script causes ad-hoc tasks with inputs and outputs described to be out of date because the actions change (presumably classpath?).

### Expected Behavior

Ad-hoc tasks should be `UP-TO-DATE` when the build script changes but the inputs/outputs do not.

### Current Behavior

Tasks are out of date, and build cache will not function as expected when the build script changes.

### Context

Iterating on a build script causes all the tasks before to be out of date when they should be. Leads to a bunch of work that should be unnecessary.

#### Kotlin:

* `build.gradle.kts`

    ```kotlin
    tasks.register("kotlinTask") {
      inputs.property("kotlinProp", 1)
      outputs.dir(layout.buildDirectory.dir("kotlinDir"))
      doFirst("empty") {}
    }
    ```

* Run `./gradlew kotlinTask --info`
* Run `./gradlew kotlinTask --info` again, see it is `UP-TO-DATE`
* Add some other code to build script that does not modify the build script. For example, change th ebuild script to:

    ```kotlin
    tasks.register("kotlinTask") {
      inputs.property("kotlinProp", 1)
      outputs.dir(layout.buildDirectory.dir("kotlinDir"))
      doFirst("empty") {}
    }
    tasks.register("other")
    ```

* Run `./gradlew kotlinTask --info`

        > Task :kotlinTask             
        Task ':kotlinTask' is not up-to-date because:                  
          Task ':kotlinTask' has additional actions that have changed

#### Groovy

The same execution strategy in Groovy will result in the task being `UP-TO-DATE`.

* `build.gradle`

    ```groovy
    tasks.register('groovyTask') {
      inputs.property('groovyProp', 1)
      outputs.dir(layout.buildDirectory.dir('groovyDir'))
      doFirst('empty') {}
    }
    ```

* Run `./gradlew groovyTask --info`
* Run `./gradlew groovyTask --info` again, see it is `UP-TO-DATE`
* Add some other code to build script that does not modify the build script. For example, change th ebuild script to:

    ```groovy
    tasks.register('groovyTask') {
      inputs.property('groovyProp', 1)
      outputs.dir(layout.buildDirectory.dir('groovyDir'))
      doFirst('empty') {}
    }
    tasks.register('other')
    ```

* Run `./gradlew groovyTask --info`, see it is `UP-TO-DATE`

### Your Environment

 * `gradle --version`

**Comment by wolfs** [comment]
That is expected behaviour: the implementation of the task and its actions are always an input to the task.

For the Groovy version the same should happen - so the task should be out-of-date when you change the build script. I don't quite understand why it is not the case.

The implementation of the task action is bound to the action's classpath - since if the classpath changes the behaviour of the action will change. Gradle's up-to-date checks and build cache key calculation take the implementation of the task and its actions into account.

See the documentation in the [build cache guide](https://guides.gradle.org/using-build-cache/#custom_actions) and the [user guide](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:how_does_it_work).

**Comment by StefMa** [comment]
> For the Groovy version the same should happen - so the task should be out-of-date when you change the build script. I don't quite understand why it is not the case.

Do I understand it correctly, that the issue for this issue is tracked in #7923? 🤔

**Comment by wolfs** [comment]
Yes, the actual issue is tracked in https://github.com/gradle/gradle/issues/7923. The problem is that the Groovy task is not out of date, everything is working well for Kotlin.

**Comment by mkobit** [comment]
Thanks @wolfs , the Groovy not behaving the same is what threw me. It would definitely be useful if ad-hoc tasks were able to be up to date, but I can understand why they are not. It definitely makes the runtime input registration API a bit more problematic and requires more use of `buildSrc` to define either custom a tasks or `Action` implementations to be used instead.
